File: src/gobby/code_index/chunker.py
"""File content chunker for full-text search indexing.

Splits file content into overlapping line-based chunks suitable for
FTS5 indexing. Each chunk captures ~100 lines with 10-line overlap
so search hits near chunk boundaries aren't lost.
"""

from __future__ import annotations

from gobby.code_index.models import ContentChunk

# Chunk parameters
CHUNK_SIZE = 100  # lines per chunk
CHUNK_OVERLAP = 10  # overlap between consecutive chunks


def chunk_file_content(
    source: bytes,
    rel_path: str,
    project_id: str,
    language: str | None = None,
    chunk_size: int = CHUNK_SIZE,
    overlap: int = CHUNK_OVERLAP,
) -> list[ContentChunk]:
    """Split file content into overlapping chunks.

    Args:
        source: Raw file bytes.
        rel_path: File path relative to project root.
        project_id: Project identifier.
        language: Language name (e.g., "python", "typescript").
        chunk_size: Lines per chunk.
        overlap: Lines of overlap between consecutive chunks.

    Returns:
        List of ContentChunk objects ready for storage.
    """
    try:
        text = source.decode("utf-8", errors="replace")
    except Exception:
        return []

    lines = text.splitlines(keepends=True)
    if not lines:
        return []

    chunks: list[ContentChunk] = []
    step = max(1, chunk_size - overlap)
    chunk_index = 0

    for start in range(0, len(lines), step):
        end = min(start + chunk_size, len(lines))
        chunk_content = "".join(lines[start:end])

        # Skip empty chunks
        if not chunk_content.strip():
            if end >= len(lines):
                break
            continue

        chunks.append(
            ContentChunk(
                id=ContentChunk.make_id(project_id, rel_path, chunk_index),
                project_id=project_id,
                file_path=rel_path,
                chunk_index=chunk_index,
                line_start=start + 1,  # 1-indexed
                line_end=end,  # 1-indexed inclusive
                content=chunk_content,
                language=language,
            )
        )
        chunk_index += 1

        if end >= len(lines):
            break

    return chunks

File: rust/gcode/src/index/chunker.rs
//! Content chunking: 100-line chunks with 10-line overlap.
//! Ports logic from src/gobby/code_index/chunker.py.

use crate::models::ContentChunk;

const CHUNK_SIZE: usize = 100;
const CHUNK_OVERLAP: usize = 10;

/// Split file content into overlapping chunks for FTS indexing.
pub fn chunk_file_content(
    source: &[u8],
    rel_path: &str,
    project_id: &str,
    language: Option<&str>,
) -> Vec<ContentChunk> {
    let text = String::from_utf8_lossy(source);
    let lines: Vec<&str> = text.split('\n').collect();
    if lines.is_empty() {
        return Vec::new();
    }

    let step = CHUNK_SIZE.saturating_sub(CHUNK_OVERLAP).max(1);
    let mut chunks = Vec::new();
    let mut chunk_index: usize = 0;
    let mut start = 0;

    while start < lines.len() {
        let end = (start + CHUNK_SIZE).min(lines.len());
        let chunk_content: String = lines[start..end].join("\n");

        if !chunk_content.trim().is_empty() {
            chunks.push(ContentChunk {
                id: ContentChunk::make_id(project_id, rel_path, chunk_index),
                project_id: project_id.to_string(),
                file_path: rel_path.to_string(),
                chunk_index,
                line_start: start + 1,
                line_end: end,
                content: chunk_content,
                language: language.unwrap_or("unknown").to_string(),
                created_at: iso_now(),
            });
            chunk_index += 1;
        }

        if end >= lines.len() {
            break;
        }
        start += step;
    }

    chunks
}

fn iso_now() -> String {
    // Minimal ISO-8601 without chrono dependency
    use std::time::SystemTime;
    let secs = SystemTime::now()
        .duration_since(SystemTime::UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs();
    format!("{secs}")
}
